utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java

```diff
@@ -36,37 +40,23 @@ public void endOperationProfile(String operationName, DatabaseQuery databaseQuer
 
     @Override
     public Object profileExecutionOfQuery(DatabaseQuery databaseQuery, Record record, AbstractSession abstractSession) {
-
-        HashMap additionalParams = new HashMap();
-        DatabaseMetaData metadata;
-        String hostname = "database";
-        try {
-            Connection connection = abstractSession.getAccessor().getConnection();
-            metadata = connection.getMetaData();
-            additionalParams.put("url", metadata.getURL());
-            additionalParams.put("user", metadata.getUserName());
-            additionalParams.put("driver_version", metadata.getDriverVersion());
-            additionalParams.put("database_type", metadata.getDatabaseProductName());
-            additionalParams.put("database_version", metadata.getDatabaseProductVersion());
-            additionalParams.put("preparation", databaseQuery.isCallQuery() ? "call" : "statement");
-            additionalParams.put("sanitized_query", StringUtils.isEmpty(databaseQuery.getSQLString()) ? "" : databaseQuery.getSQLString());
-
-            try {
-                hostname = new URI((new URI(metadata.getURL())).getSchemeSpecificPart()).getHost();
-                hostname = connection.getCatalog() + "@" + hostname;
-            } catch (URISyntaxException exception) {
-                logger.warn("Error parsing database host name.");
+        if (recorder.getCurrentSegmentOptional().isPresent()) {
+            final Optional<Connection> maybeConnection = Optional.ofNullable(abstractSession.getAccessor()).map(Accessor::getConnection);
+            Map<String, Object> databaseMetadata = new HashMap<>();
+            String hostname = "database";
+            if (maybeConnection.isPresent()) {
+                final Connection connection = maybeConnection.get();
+                databaseMetadata = getMetadata(databaseQuery, connection);
+                hostname = getDatabaseHostName(connection);
+            } else {
+                logger.warn("There is no database connection available");
             }
-        } catch (SQLException exception) {
-            logger.warn("Error getting database connection details.");
-        }
 
-        if (recorder.getCurrentSegmentOptional().isPresent()) {
             Subsegment subsegment = recorder.beginSubsegment(hostname);
             subsegment.putMetadata("monitor_name", databaseQuery.getMonitorName());
             subsegment.putMetadata("calling_class", databaseQuery.getClass().getSimpleName());
             subsegment.setNamespace(Namespace.REMOTE.toString());
-            subsegment.putAllSql(additionalParams);
+            subsegment.putAllSql(databaseMetadata);
```

Comment:
This looks a lot nicer but it’s a shame that things like the `"database"` default and the empty database metadata map appear in two places.

We could do something like this?

```java
if (recorder.getCurrentSegmentOptional().isPresent()) {
    final Optional<Connection> maybeConnection = Optional.ofNullable(abstractSession.getAccessor()).map(Accessor::getConnection);
    
    if (!maybeConnection.isPresent()) {
        logger.warn("There is no database connection available");
    }

    Map<String, Object> databaseMetadata = maybeConnection.map(connection -> getMetadata(databaseQuery, connection)).orElseGet(HashMap::new);
    String hostname = maybeConnection.flatMap(this::getDatabaseHostName).orElse("database");

    Subsegment subsegment = recorder.beginSubsegment(hostname);
    subsegment.putMetadata("monitor_name", databaseQuery.getMonitorName());
    subsegment.putMetadata("calling_class", databaseQuery.getClass().getSimpleName());
    subsegment.setNamespace(Namespace.REMOTE.toString());
    subsegment.putAllSql(databaseMetadata);

    try {
        return abstractSession.internalExecuteQuery(databaseQuery, (AbstractRecord) record);
    } finally {
        subsegment.end();
    }
} else {
    logger.info("No segment found, executing query without recording subsegment data.");
    return abstractSession.internalExecuteQuery(databaseQuery, (AbstractRecord) record);
}
```

This would require following the earlier suggestion to make `getDatabaseHostName(Connection)` return an `Optional<String>`.